### PR TITLE
Stdification idiom

### PR DESCRIPTION
**Pull request by darnuria**
It's a big chunk I tried to get every commit compiling standalone if any commit is better in a separated PR tell me. ;)

NOTE: It's done as a FLOSS dev, not on work hours. no problem with that always good to improve libre/free software.

Testing it in real setting is needed but I didn't touched too much at the IO-level.

EDIT: I am splicing it, will test separately each one.

### CODE REVIEW

File: src/jbod/disks.rs [hunk 1]
```diff
@@ -258,28 +251,26 @@ pub mod DiskShelf {
...
+            led_fault_path
         } else {
-            return "NONE".to_string();
+            "NONE".to_string()
         }
     }
 
     /// Here we write 0 or 1 into the disk led file
-    fn set_disk_led_locate(disk: String, option: &str) {
+    fn set_disk_led_locate(disk: &str, option: &str) {
```
Comment (darnuria) [inline src/jbod/disks.rs:264]: Using `AsRef<str>` would have been a solution but it's a bit of a burden.

https://www.philipdaniels.com/blog/2019/rust-api-design/

File: src/jbod/disks.rs [hunk 2]
```diff
@@ -510,7 +510,7 @@ pub mod DiskShelf {
     ///
     /// * `options` - a reference of ArgMatches
     ///
-    pub fn jbod_led_switch(options: &ArgMatches) -> Result<(), ()> {
+    pub fn jbod_led_switch(options: &ArgMatches) -> Result<(), Box<dyn std::error::Error>> {
```
Comment (darnuria) [inline src/jbod/disks.rs:519]: Trait Object Boxing errors is pretty standard on application, there is some good crate about that like [anyhow](https://github.com/dtolnay/anyhow)

File: src/jbod/disks.rs [hunk 3]
```diff
@@ -527,28 +527,22 @@ pub mod DiskShelf {
...
-                .unwrap_or(&"/dev/null".to_string())
-                .to_string();
+            let disk = options.value_of("fault").unwrap_or("/dev/null");
             if on {
-                set_disk_led_fault(disk.clone(), &"1".to_string());
+                set_disk_led_fault(disk, "1");
             }
             if off {
-                set_disk_led_fault(disk.clone(), &"0".to_string());
+                set_disk_led_fault(disk, "0");
```
Comment (darnuria) [inline src/jbod/disks.rs:551]: Ended up with a solution to avoid `String` conversion.

File: src/main.rs [hunk 4]
```diff
@@ -81,31 +81,28 @@ fn enclosure_overview(option: &ArgMatches) -> Result<(), ()> {
...
                     print!(" Disk: {:<10}", disk.device_path.green(),);
-                    if disk.device_map == "NONE" {
-                        print!(" Map: {:<10}", disk.device_map.yellow());
+                    let device_map_color = if disk.device_map == "NONE" {
+                        disk.device_map.yellow()
                     } else {
-                        print!(" Map: {:<10}", disk.device_map.green());
-                    }
+                        disk.device_map.green()
+                    };
```
Comment (darnuria) [inline src/main.rs:88]: if-expr are neats, produce quite good assembly and nice to read it comes from Ocaml/Haskell/Scala inspiration.

File: src/main.rs [hunk 5]
```diff
@@ -81,31 +81,28 @@ fn enclosure_overview(option: &ArgMatches) -> Result<(), ()> {
...
                     print!(" Model: {:<10}", disk.model.blue());
                     print!(" Serial: {:<10} ", disk.serial.blue());
-                    let temp_conv = disk.temperature.parse::<i32>().unwrap();
-                    if temp_conv > 45 && temp_conv <= 50 {
-                        print!(
-                            "Temp: {}{:<2}",
-                            disk.temperature.yellow().bold(),
-                            "c".yellow().bold()
-                        );
+                    let temp_conv = disk.temperature.parse::<i32>()?;
```
Comment (darnuria) [inline src/main.rs:94]: `?` is really the goto way to manage `Result` and `Option` for early failling.

File: src/main.rs [hunk 6]
```diff
@@ -135,46 +132,36 @@ fn enclosure_overview(option: &ArgMatches) -> Result<(), ()> {
...
-    if let Some(port) = option.value_of("port") {
-        default_port = port;
-    }
-
-    if let Some(ip) = option.value_of("ip-address") {
-        default_address = ip;
-    }
+fn fork_prometheus(option: &ArgMatches) -> Result<(), Box<dyn std::error::Error>> {
+    let default_port = option.value_of("port").unwrap_or("9945");
+    let default_address = option.value_of("ip-address").unwrap_or("0.0.0.0");
```
Comment (darnuria) [inline src/main.rs]: Great spot for a `unwrap_or` all

File: src/main.rs [hunk 7]
```diff
@@ -135,46 +132,36 @@ fn enclosure_overview(option: &ArgMatches) -> Result<(), ()> {
...
         Ok(ForkResult::Child) => {
             Command::new(Util::JBOD_EXPORTER)
                 .args(&[default_address, default_port])
-                .spawn()
-                .expect("Failed to spawn the target process");
+                .spawn()?;
             exit(0);
         }
-        Err(_) => println!("Fork Failed"),
+        Err(e) => Err(Box::new(e)),
```
Comment (darnuria) [inline src/main.rs:170]: Boxing error allows merging error together if the Error implements `std::error::Error` would not do that in a lib but in cli-app it's the way to go. A more powerful way in an app is anyhow like said above.

https://doc.rust-lang.org/rust-by-example/error/multiple_error_types/boxing_errors.html

File: src/main.rs [hunk 8]
```diff
@@ -268,5 +255,6 @@ fn main() {
         ("led", Some(m)) => DiskShelf::jbod_led_switch(m),
         ("prometheus", Some(m)) => fork_prometheus(m),
         _ => Ok(help()),
-    };
+    }
+    .expect("jbod encountered an internal issue:")
```
Comment (darnuria) [inline src/main.rs:266]: Would also print the `Err`.

Comment (darnuria) [reply src/main.rs:266]: Something like: 
[playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=use%20std%3A%3Aio%3A%3AWrite%3B%0Afn%20main()%20%7B%0A%20%20%20%20%22toto%22.parse%3A%3A%3Ci32%3E().expect(%22%3CInternal%20Error%3E%22)%3B%0A%7D)
```rust
fn main() {
    "toto".parse::<i32>().expect("<Internal Error>");
}
```
```
thread 'main' panicked at '<Internal Error>: ParseIntError { kind: InvalidDigit }', src/main.rs:3:27
note: [run with `RUST_BACKTRACE=1` environment variable to display a backtrace](https://play.rust-lang.org/#)
```

File: src/jbod/disks.rs [hunk 9]
```diff
@@ -209,13 +209,10 @@ pub mod DiskShelf {
...
             for line in stdout_lines {
-                let line_unwrap = line.unwrap().to_owned();
+                let line_unwrap = line.unwrap();
                 let output_split: Vec<&str> = line_unwrap.split_whitespace().collect();
-                if output_split.len() > 1 {
-                    disks.insert(output_split[0].to_string(), output_split[1].to_string());
-                } else {
-                    disks.insert(output_split[0].to_string(), "NONE".to_string());
-                }
+                let val = output_split.get(1).unwrap_or(&"NONE");
```
Comment (darnuria) [inline src/jbod/disks.rs:214]: [Vec::get](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get) and [Vec::get_mut](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get_mut) are great, rustc generate code without panic-bound-check-error code with it.

Code end up quite well optimized and it's quite readable.

The whole futue API who overs `get` and `get_mut` is https://doc.rust-lang.org/std/slice/trait.SliceIndex.html but for now it's only stable on `Vec<T>`.

**Review by darnuria — COMMENTED** [review]
Time to sleep, happy to read question and review remark later!

**Comment by araujobsd** [comment]
This patch breaks the disk list, need some more investigation.
```
./jbod list -d
 SLOT      | DEVICE    | VENDOR | MODEL  | REVISION | SERIAL
-----------+-----------+--------+--------+----------+------------------
 11:0:14:0 | /dev/sg16 | DP     | BP15G+ | 3.04     | 32CEA7F073112100
     '
root@filerlab:/tmp#
```

**Comment by darnuria** [comment]
For sure, I will split this PR and write some test on the function who don't do side effect on spare time. Will reduce the surface of full checks. Also will take some time to setup jbod it's interesting have a lot of olds disk at home.

**Comment by araujobsd** [comment]
I'm closing this PR as I had no return from the submitter

**Pull request closed without merging**